File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/HandheldController/SleepMode/SleepModeViewController.m
//
//  SleepModeViewController.m
//  DJISdkDemo
//
//  Copyright © 2015 DJI. All rights reserved.
//
/**
 *  This file demonstrates how to set the power mode of a handheld controller. For more information
 *  about the power mode, please refer to the inline documentation for DJIHandheldPowerMode. 
 *
 *  CAUTION: user can turn off the handheld device through SDK. However, once the device is turned 
 *  off, user need to turn on the device using the physical button. 
 */
#import "DemoUtility.h"
#import "SleepModeViewController.h"

@interface SleepModeViewController () <DJIHandheldControllerDelegate>

@property (weak, nonatomic) IBOutlet UIButton *sleepButton;
@property (weak, nonatomic) IBOutlet UIButton *awakeButton;
@property (weak, nonatomic) IBOutlet UIButton *shutdownButton;

@end

@implementation SleepModeViewController

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld) {
        [handheld setDelegate:self];
    }
    else {
        ShowResult(@"There is no handheld controller. ");
        [self.sleepButton setEnabled:NO];
        [self.awakeButton setEnabled:NO];
        [self.shutdownButton setEnabled:NO];
    }
}

-(void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld && handheld.delegate == self) {
        [handheld setDelegate:nil];
    }
}

- (IBAction)onSleepButtonClicked:(id)sender {
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld) {
        [self.sleepButton setEnabled:NO];
        [self sendPowerMode:DJIHandheldPowerModeSleeping];
    }
}

- (IBAction)onAwakeButtonClicked:(id)sender {
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld) {
        [self.awakeButton setEnabled:NO];
        [self sendPowerMode:DJIHandheldPowerModeOn];
    }
}

- (IBAction)onShutdownButtonClicked:(id)sender {
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld) {
        [self.shutdownButton setEnabled:NO];
        [self sendPowerMode:DJIHandheldPowerModeOff];
    }
}

-(void) sendPowerMode:(DJIHandheldPowerMode)mode {
    DJIHandheldController* handheld = [DemoComponentHelper fetchHandheldController];
    if (handheld) {
        [handheld setPowerMode:mode withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"ERROR: setHandheldPowerMode failed. %@", error.description);
            }
            else {
                ShowResult(@"SUCCESS: setHandheldPowerMode. ");
            }
        }];
    }
}

#pragma mark - DJIHandheldControllerDelegate
-(void)handheldController:(DJIHandheldController *)controller didUpdatePowerMode:(DJIHandheldPowerMode)powerMode {
    switch (powerMode) {
        case DJIHandheldPowerModeSleeping:
            [self.sleepButton setEnabled:NO];
            [self.awakeButton setEnabled:YES];
            [self.shutdownButton setEnabled:YES];
            break;

        case DJIHandheldPowerModeOn:
            [self.sleepButton setEnabled:YES];
            [self.awakeButton setEnabled:NO];
            [self.shutdownButton setEnabled:YES];
            break;
            
        case DJIHandheldPowerModeOff:
            [self.sleepButton setEnabled:NO];
            [self.awakeButton setEnabled:NO];
            [self.shutdownButton setEnabled:NO];
            break;
            
        default:
            break;
    }
}


@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/HandheldController/SleepMode/SleepModeViewController.swift
//
//  SleepModeViewController.swift
//  DJISDKSwiftDemo
//
//  Copyright © 2015 DJI. All rights reserved.
//
import DJISDK

class SleepModeViewController: DJIBaseViewController, DJIHandheldControllerDelegate {
    @IBOutlet weak var sleepButton: UIButton!
    @IBOutlet weak var awakeButton: UIButton!
    @IBOutlet weak var shutdownButton: UIButton!
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld != nil {
            handheld!.delegate = self
        }
        else {
            self.showAlertResult("There is no handheld controller. ")
            self.sleepButton.isEnabled = false
            self.awakeButton.isEnabled = false
            self.shutdownButton.isEnabled = false
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld?.delegate === self {
            handheld!.delegate = nil
        }
    }
    
    @IBAction func onSleepButtonClicked(_ sender: AnyObject) {
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld != nil {
            self.sleepButton.isEnabled = false
            self.sendPowerMode(DJIHandheldPowerMode.sleeping)
        }
    }
    
    @IBAction func onAwakeButtonClicked(_ sender: AnyObject) {
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld != nil {
            self.awakeButton.isEnabled = false
            self.sendPowerMode(DJIHandheldPowerMode.awake)
        }
    }
    
    @IBAction func onShutdownButtonClicked(_ sender: AnyObject) {
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld != nil {
            self.shutdownButton.isEnabled = false
            self.sendPowerMode(DJIHandheldPowerMode.powerOff)
        }
    }
    
    func sendPowerMode(_ mode: DJIHandheldPowerMode) {
        let handheld: DJIHandheldController? = self.fetchHandheldController()
        if handheld != nil {
            handheld!.setHandheldPowerMode(mode, withCompletion: {[weak self](error: Error?) -> Void in
                if error != nil {
                    self?.showAlertResult("ERROR: setHandheldPowerMode failed: \(error!)")
                }
                else {
                    self?.showAlertResult("SUCCESS: setHandheldPowerMode. ")
                }
            })
        }
    }
    
    func handheldController(_ controller: DJIHandheldController, didUpdate powerMode: DJIHandheldPowerMode) {
        switch powerMode {
        case DJIHandheldPowerMode.sleeping:
            self.sleepButton.isEnabled = false
            self.awakeButton.isEnabled = true
            self.shutdownButton.isEnabled = true
        case DJIHandheldPowerMode.awake:
            self.sleepButton.isEnabled = true
            self.awakeButton.isEnabled = false
            self.shutdownButton.isEnabled = true
        case DJIHandheldPowerMode.powerOff:
            self.sleepButton.isEnabled = false
            self.awakeButton.isEnabled = false
            self.shutdownButton.isEnabled = false
        default:
            break
        }
        
    }
}
